test/cloudwatchlogs/publish_logs_test.go

```diff
@@ -172,6 +173,85 @@ func TestRotatingLogsDoesNotSkipLines(t *testing.T) {
 	assert.True(t, ok)
 }
 
+// TestCloudWatchLogsBuffer writes N number of logs, and then validates that N logs
+// are queryable from CloudWatch Logs
+// and buffer size works
+// buffer size is 1mb
+func TestCloudWatchLogsBuffer(t *testing.T) {
+	// this uses the {instance_id} placeholder in the agent configuration,
+	// so we need to determine the host's instance ID for validation
+	instanceId := awsservice.GetInstanceId()
+	log.Printf("Found instance id %s", instanceId)
+	common.DeleteFile(common.AgentLogFile)
+	common.TouchFile(common.AgentLogFile)
+	start := time.Now()
+
+	numberOfLogsPerFile := 1000
+	numberOfLogFiles := 100
+	filePathFormat := "/tmp/logs_buffer/"
+	fileName := filePathFormat + "test.log"
+
+	for i := 0; i < numberOfLogFiles; i++ {
+		fileDir := filePathFormat
+		filePath := fileName + "." + strconv.Itoa(i)
+		defer awsservice.DeleteLogGroupAndStream(instanceId, filePath)
+		err := os.MkdirAll(fileDir, os.ModePerm)
+		if err != nil {
+			t.Fatalf("Error occurred creating log dir for writing: %v", err)
+		}
+		f, err := os.Create(filePath)
+		if err != nil {
+			t.Fatalf("Error occurred creating log file for writing: %v", err)
+		}
+		defer f.Close()
+		defer os.Remove(filePath)
+		go writeLogs(t, f, numberOfLogsPerFile)
+	}
+
+	common.CopyFile("resources/config_log_buffer.json", configOutputPath)
+	common.StartAgent(configOutputPath, true, false)
+	time.Sleep(time.Minute * 2)
+	common.StopAgent()
+
+	agentLog, err := common.RunCommand(common.CatCommand + common.AgentLogFile)
+	if err != nil {
+		return
+	}
+	t.Logf("Agent logs %s", agentLog)
+	// confirm buffer is hit
+	// confirm blocking of next log search tick
+	// confirm blocking of reading next line
+	// confirm buffer is cleared at the end
+	assert.True(t, strings.Contains(agentLog, "blocking adding new files for one second"))
```

Comment:
Looking for log lines is probs a bad idea. I am not sure how to test this otherwise.